### PR TITLE
Refactor new date picker

### PR DESCRIPTION
**Pull request by ArekChr**
### Details
This pull request introduces significant improvements to the Date Picker component by refactoring the codebase and implementing logic changes. The main objectives are to enhance reusability by removing navigation.

### Fixed Issues
$ https://github.com/Expensify/App/issues/20630
PROPOSAL: https://github.com/Expensify/App/issues/20630#issuecomment-1597567254

### Tests
1. Go to Date of Birth in Personal Details
2. Select the year (nothing changes in the input)
3. Select month (nothing changes in the input)
4. Click on the day (the date is changing in the input)
5. Change the month using right-left arrows (nothing changes in the input)
6. Click on the day (the date is changing in the input)

- [x] Verify that no errors appear in the JS console

### Offline tests
No offline tests needed.

### QA Steps
1. Go to Date of Birth in Personal Details
2. Select the year (nothing changes in the input)
3. Select month (nothing changes in the input)
4. Click on the day (the date is changing in the input)
5. Change the month using right-left arrows (nothing changes in the input)
6. Click on the day (the date is changing in the input)

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/24796318/2905b43c-a216-41a7-9d90-7522a22d3c42

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/24796318/c4a646ef-7435-46d7-9590-14361c1b3550

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/24796318/daef224d-66b7-4ddf-8a5a-e9b863950c97

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/24796318/ef4d914b-15b1-4d29-a40e-e065badabd32

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/24796318/3a3bb5ff-d401-4474-a950-c46982101290

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/24796318/10aa39da-71f5-4669-b648-d7ae2c74890e

</details>

### CODE REVIEW

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (fabioh8010) [inline src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:27]: Why not use `windowWidth` from `withWindowDimensions` hook?

File: src/components/NewDatePicker/CalendarPicker/index.js
Comment (fabioh8010) [inline src/components/NewDatePicker/CalendarPicker/index.js:21]: This spread from the HOC should be at the bottom.

File: src/components/NewDatePicker/CalendarPicker/index.js
Comment (fabioh8010) [inline src/components/NewDatePicker/CalendarPicker/index.js:94]:
```suggestion
        this.setState((prev) => ({currentDateView: moment(prev.currentDateView).subtract(1, 'months').toDate()}));
```

Maybe `months` to be more concise with what you did on `moment(prev.currentDateView).set('year', year).toDate()` for example?

File: src/components/NewDatePicker/CalendarPicker/index.js
Comment (fabioh8010) [inline src/components/NewDatePicker/CalendarPicker/index.js:101]:
```suggestion
        this.setState((prev) => ({currentDateView: moment(prev.currentDateView).add(1, 'months').toDate()}));
```

Same.

File: src/components/NewDatePicker/index.js
Comment (fabioh8010) [inline src/components/NewDatePicker/index.js:14]: This spread from the HOC should be at the bottom.

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (fabioh8010) [inline src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:60]:
```suggestion
        <Portal hostName="RightModalNavigator">
```

File: src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js
Comment (fabioh8010) [inline src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js:10]:
```suggestion
function RightModalNavigator() {
```

File: src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js
Comment (fabioh8010) [inline src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js:110]:
```suggestion
            <PortalHost name="RightModalNavigator" />
```

File: src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js
Comment (fabioh8010) [inline src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js:77]:
```suggestion
                    minDate={moment().subtract(CONST.DATE_BIRTH.MAX_AGE, 'years').toDate()}
                    maxDate={moment().subtract(CONST.DATE_BIRTH.MIN_AGE, 'years').toDate()}
```

Maybe `years ` to be more concise with previous usages in other files?

**Review by fabioh8010 — COMMENTED** [review]
Just nitpicks

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:60]: The typo comes from RigthModalNavigator and will be fixed in proper PR, it is the same so helps with searching components

File: src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js
Comment (ArekChr) [reply src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.js:10]: This is out of scope

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:27]: Good catch!

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js]: It can, but for now, only one component uses it. 

Maybe rename the `CalendarPicker` dir to `components` inside the `NewDatePicker` dir? WDYT?

Comment (gedu) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js]: no, I think it is ok

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (thiagobrez) [inline src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:43]: What happens if you remove the `eslint-disable` and put `translateX` in the dependency array? does it cause an infinite loop?

I'm not too familiar with Reanimated, but thought `useSharedValue` returned a ref, so it could be safe to add in the deps array

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (thiagobrez) [inline src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:52]: Same here, ideally if we could remove the `eslint-disable` would be nice

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:26]: It's not correct. Good catch! fixing

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:43]: @thiagobrez It doesn't cause an infinite loop. I generally prefer not to include unnecessary variables in the dependency array. I like to keep a clear understanding of which variables an effect depends on.

Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:43]: I'll seek out an ESLint configuration that can ignore such references.

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (burczu) [inline src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:56]:
```suggestion
    if (!visible) {
        return null;
    }
```

and I would even like to force this with linter ;)

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:43]: I understand, but if it causes no harm in adding it to the deps array, I would include it. Otherwise in the future, when someone is refactoring this, it can lead to bugs we are not aware of right now. Also, if it's a stable reference, it won't trigger a re-render anyway

Comment (ArekChr) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:43]: Sure, let's add that then

**Comment by mountiny** [comment]
@fedirjh Can you reproduce the issues with animations for the year picker which Arek mentioned? SWM was not able to repro

Also Arek is ooo so I have bumped Callstack and someone else will pick this up from tomorrow. @fedirjh lets please priotize this review and testing so we can ideally get this merge by EOW 🙇  thank you 🙇

**Comment by mountiny** [comment]
Thanks for confirming hopefully we wont be blocked on hat then

**Comment by burczu** [comment]
Hey @fedirjh! Regarding this:

> Hey there @ArekChr , We have some bugs :
> 
> * When Selecting a Year or a Month, it doesn’t update the date
> 

It's intended - please check this slack thread: https://expensify.slack.com/archives/C049HHMV9SM/p1687425911714009

**Review by fedirjh — COMMENTED** [review]
Overall code looks good, I have only one concern raised [here](https://github.com/Expensify/App/pull/21792/files#r1252183953)

File: src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/ScreenSlideAnimation.js:1]: Great idea! What you proposed works flawlessly, with the only difference being that the modal has a backdrop, that the YearPickerPage did not have previously. But this matches the AvatarCropModal and brings more design consistency, so I think it's a even bigger win


https://github.com/Expensify/App/assets/26878038/62bc5e72-0751-4fae-bc4d-9120c815c12c

**Comment by thiagobrez** [comment]
>Overall code looks good, I have only one concern raised [here](https://github.com/Expensify/App/pull/21792/files#r1252183953)

@fedirjh's suggestion works great. So we can stop worrying about the animation issues since we're ditching `ScreenSlideAnimation` in favor of `Modal` (I couldn't reproduce the issue anyway). Left a video of the Modal implementation in the [thread above](https://github.com/Expensify/App/pull/21792#discussion_r1254483905)

>We are missing prop definition for this component.

>Little prop fix

>Same here

Fixed all propTypes pointed out

>Extra horizontal margin for the years selector

Fixed. Also, in [Phase 2](https://github.com/Expensify/App/issues/20353) of the [Selection Lists refactor](https://github.com/Expensify/App/issues/11795) (which I'll continue working right after we get this one merged), I'll be replacing `OptionsSelector` here for `SelectionList`, the new standardised component for selection lists

**Comment by thiagobrez** [comment]
>Text selected upon clicking arrows or calendar icon

Working on this last remaining issue ⬆️

**Comment by thiagobrez** [comment]
I believe I have everything ready now. We're contacting Arek to get me permissions to push to his branch, but if we can't until tomorrow morning I'll raise a new PR

**Comment by mountiny** [comment]
@thiagobrez love this, thank you 🙇

**Comment by thiagobrez** [comment]
@fedirjh Pushed the fixes, let me know what you think 🙌🏻

**Review by fedirjh — COMMENTED** [review]
Looks like we have some performance issue with IOS , the transition does not looks smooth, It seems that the year list is updated during the transition 


https://github.com/Expensify/App/assets/36869046/a34ac973-1c95-4f97-b873-45139d9133be

File: src/components/NewDatePicker/CalendarPicker/YearPickerPage.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerPage.js]: Not sure, but I removed the whole `ScreenWrapper` anyway as it is not meant to be used inside Modals. This also fixed the blank space above the header

File: src/components/NewDatePicker/CalendarPicker/YearPickerPage.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerPage.js]: Thanks for pointing out, I am not super familiar with `moment` and thought that was returning a number

File: src/components/NewDatePicker/CalendarPicker/index.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/index.js]: Good call, done

File: src/components/NewDatePicker/CalendarPicker/YearPickerPage.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerPage.js]: Fixed by removing `ScreenWrapper`

**Comment by mountiny** [comment]
thanks for the reviews and testing so far. great job, lets egt this done on Monday 🎉

**Review by fedirjh — COMMENTED** [review]
Looks we are very close, left few comments.

File: src/components/NewDatePicker/CalendarPicker/YearPickerModal.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerModal.js:57]: Yep, `Modal` spreads the props:

```jsx
function Modal(props) {
    return (
        <BaseModal
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
        >
            {props.children}
        </BaseModal>
    );
}
```

and then it goes all the way down to `BaseModal`, which uses it:

```jsx
<ReactNativeModal
    hideModalContentWhileAnimating={this.props.hideModalContentWhileAnimating}
    ...
>
    ...
</ReactNativeModal>
```

`hideModalContentWhileAnimating` comes from `react-native-modal` and is used to prevent flashes in the UI when `useNativeDriver` is true. 

See here:
https://github.com/react-native-modal/react-native-modal/issues/343#issuecomment-520243962

and here:
https://github.com/react-native-modal/react-native-modal/pull/116

Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerModal.js:57]: @fedirjh Yeah, seems like there are a few other props missing too. I could add them as part of this PR, do you think it makes sense?

File: src/components/NewDatePicker/CalendarPicker/YearPickerModal.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerModal.js]: I wasn't aware, thanks for this!

File: src/components/NewDatePicker/CalendarPicker/YearPickerModal.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerModal.js]: Cool, I like this approach. But that does not reset the search alone, because the modal is never unmounted, it just becomes invisible. So I also added an effect to clear the search text when it closes

**Comment by thiagobrez** [comment]
@fedirjh Sent fixes to all your latest comments! :)

**Comment by thiagobrez** [comment]
I see that some unit tests failed and I'm working on it 👍🏻

**Comment by thiagobrez** [comment]
Fixed 👌🏻

**Review by fedirjh — COMMENTED** [review]
Code looks good to me. Left one small comment.

File: src/components/NewDatePicker/CalendarPicker/YearPickerModal.js
Comment (thiagobrez) [reply src/components/NewDatePicker/CalendarPicker/YearPickerModal.js]: Nice, done

**Review by fedirjh — APPROVED** [review]
Looks good to me.

**Comment by mountiny** [comment]
@fedirjh Great job on the review and thanks @thiagobrez for acting fast!

**Review by mountiny — APPROVED** [review]
Thanks everyone, great job on this one 🎉

**Comment by s77rt** [comment]
Awesome work! Love the simplification.

File: src/components/NewDatePicker/CalendarPicker/index.js
Comment (mollfpr) [inline src/components/NewDatePicker/CalendarPicker/index.js:112]: ✋ Coming from https://github.com/Expensify/App/issues/33006

The refactor for `moveToPrevMonth` and `moveToNextMonth` doesn't update the year.

File: src/components/NewDatePicker/CalendarPicker/YearPickerModal.js
Comment (hungvu193) [inline src/components/NewDatePicker/CalendarPicker/YearPickerModal.js:56]: We should include `shouldHandleNavigationBack` props to handle physical back button on mWeb. More details in:
https://github.com/Expensify/App/issues/52383#issuecomment-2473437159

**Pull request merged**
